File: Nachhilfe 29.06.2012/JavAnt_Vorlage/Useful/Network.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Net.Sockets;

namespace Useful
{
    public class Network
    {
        public string receive(TcpClient client)
        {
            try
            {
                var stream = client.GetStream();
                byte[] buffer = new byte[4];
                var readBytes = stream.Read(buffer, 0, 4);
                if (readBytes == 0)
                {
                    throw new Exception("connection to server lost");
                }
                else
                {
                    int msgSize = BitConverter.ToInt32(buffer, 0);
                    buffer = new byte[msgSize];
                    readBytes = stream.Read(buffer, 0, msgSize);
                    if (readBytes == 0)
                        throw new Exception("connection to server lost");
                    else
                    {
                        var msg = System.Text.Encoding.Unicode.GetString(buffer);
                        return msg;
                    }
                }
            }
            catch (Exception e)
            {

            }
            return "";
        }

        public void send(TcpClient client, string msg)
        {
            //printfn "sent %s" msg
            var stream = client.GetStream();
            var msgBuffer = Encoding.Unicode.GetBytes(msg);
            var buffer = BitConverter.GetBytes(msgBuffer.Length);
            stream.Write(buffer, 0, 4);
            stream.Flush();
            stream.Write(msgBuffer, 0, msgBuffer.Length);
            stream.Flush();
        }
    }
}

File: Nachhilfe 29.06.2012/JavAnt_Vorlage/Useful/Network.fs
// Learn more about F# at http://fsharp.net

namespace Useful

module Network =
    open System
    open System.Net
    open System.Net.Sockets
    open System.Text
    open System.IO
    open System.Threading

    // creates a server
    let CreateServer (hostname:string) (port:int) =
        let s = new TcpListener(IPAddress.Parse(hostname), port)
        do s.Start()
        s

    // creates a client
    let CreateClient (hostname:string) (port:int) =
        new TcpClient(hostname, port)

    // create a client on localhost with a standard port
    let CreateLocalClient(port) =
        CreateClient "127.0.0.1" port

    // create a server on localhost with a standard port
    let CreateLocalServer(port) =
        CreateServer "127.0.0.1" port

    // sends a string as well as its length
    let send (client:TcpClient, msg:string) =
        //printfn "sent %s" msg
        let stream = client.GetStream()
        let msgBuffer = Encoding.Unicode.GetBytes(msg)
        let buffer = BitConverter.GetBytes(msgBuffer.Length)
        do stream.Write(buffer, 0, 4)
        do stream.Flush()
        do stream.Write(msgBuffer, 0, msgBuffer.Length)
        do stream.Flush()

    // receives a string as well as its length
    let receive (client:TcpClient) =
        try
            let stream = client.GetStream()
            let buffer = Array.zeroCreate 4
            let readBytes = stream.Read(buffer , 0 , 4)
            if readBytes = 0 then
                failwith "Client disconnected"
            else
                let msgSize = BitConverter.ToInt32(buffer , 0)
                let buffer = Array.zeroCreate msgSize
                let readBytes = stream.Read(buffer, 0, msgSize)
                if readBytes = 0 then failwith "Client disconnected"
                else
                    let msg = System.Text.Encoding.Unicode.GetString(buffer)
                   // printfn "received %s" msg
                    msg
        with
            | :? System.IO.IOException -> failwith "Client disconnected"

    let handleRequests (server:TcpListener) callback =
        let worker = new Thread(new ThreadStart(fun() ->
                                while Thread.CurrentThread.IsAlive do
                                    try callback(receive (server.AcceptTcpClient()))
                                    with e -> ()))
        do worker.Start()
        worker